File: client/src/meters/transition.ts
import type { MeterCatalog, MeterDefinition } from "./catalog";
import { METER_MAX_VALUE, METER_MIN_VALUE } from "./catalog";

export const MILLIS_PER_HOUR = 3_600_000 as const;
export const MAXIMUM_ATTENDED_STEP = 86_400_000 as const;

export interface MeterState {
  readonly values: Record<string, number>;
  readonly decayRemainders: Record<string, number>;
  readonly inputRemainders: Record<string, number>;
}

export interface MeterAdvanceContext {
  readonly attendedMs: number;
  readonly newFactKinds: ReadonlySet<string>;
  readonly activeContributions: ReadonlySet<string>;
}

export interface MeterBandChange {
  readonly meterId: string;
  readonly fromBand: string;
  readonly toBand: string;
  readonly direction: "up" | "down";
  readonly valueBefore: number;
  readonly valueAfter: number;
}

export function inputRemainderKey(meterId: string, inputIndex: number): string {
  return `${meterId}:${inputIndex}`;
}

export function contributionKey(slot: string, sourceId: string): string {
  return `${slot}\0${sourceId}`;
}

export function newMeterState(catalog: MeterCatalog): MeterState {
  const state: MeterState = { values: {}, decayRemainders: {}, inputRemainders: {} };
  for (const meter of catalog.meters) {
    state.values[meter.id] = meter.initialValue;
    state.decayRemainders[meter.id] = 0;
    meter.inputs.forEach((input, index) => {
      if (input.kind === "contribution_slot") state.inputRemainders[inputRemainderKey(meter.id, index)] = 0;
    });
  }
  return state;
}

export function validateMeterState(catalog: MeterCatalog, state: MeterState): void {
  const expectedValues = catalog.meters.map((meter) => meter.id).sort();
  if (!sameStrings(Object.keys(state.values).sort(), expectedValues) || !sameStrings(Object.keys(state.decayRemainders).sort(), expectedValues)) invalidState();
  const expectedInputs: string[] = [];
  for (const meter of catalog.meters) {
    if (!integerInRange(state.values[meter.id], METER_MIN_VALUE, METER_MAX_VALUE) || !integerInRange(state.decayRemainders[meter.id], 0, MILLIS_PER_HOUR - 1)) invalidState();
    meter.inputs.forEach((input, index) => {
      if (input.kind !== "contribution_slot") return;
      const key = inputRemainderKey(meter.id, index);
      expectedInputs.push(key);
      if (!integerInRange(state.inputRemainders[key], 0, MILLIS_PER_HOUR - 1)) invalidState();
    });
  }
  if (!sameStrings(Object.keys(state.inputRemainders).sort(), expectedInputs.sort())) invalidState();
}

export function advanceMeters(catalog: MeterCatalog, state: MeterState, context: MeterAdvanceContext): readonly MeterBandChange[] {
  if (!integerInRange(context.attendedMs, 0, MAXIMUM_ATTENDED_STEP) || !(context.newFactKinds instanceof Set) || !(context.activeContributions instanceof Set)) invalidState();
  validateMeterState(catalog, state);
  const changes: MeterBandChange[] = [];
  for (const meter of catalog.meters) {
    const before = state.values[meter.id]!;
    let value = before;
    if (meter.decay !== null && value === meter.decay.towardValue) {
      state.decayRemainders[meter.id] = 0;
    } else if (meter.decay !== null && context.attendedMs > 0) {
      const [steps, remainder] = wholeSteps(meter.decay.ratePerHour, context.attendedMs, state.decayRemainders[meter.id]!);
      if (value < meter.decay.towardValue) {
        value += steps;
        if (value >= meter.decay.towardValue) {
          value = meter.decay.towardValue;
          state.decayRemainders[meter.id] = 0;
        } else state.decayRemainders[meter.id] = remainder;
      } else {
        value -= steps;
        if (value <= meter.decay.towardValue) {
          value = meter.decay.towardValue;
          state.decayRemainders[meter.id] = 0;
        } else state.decayRemainders[meter.id] = remainder;
      }
    }

    let factDelta = 0;
    let rateDelta = 0;
    meter.inputs.forEach((input, inputIndex) => {
      if (input.kind === "ledger_fact") {
        if (context.newFactKinds.has(input.factKind)) factDelta += input.delta;
        return;
      }
      if (!context.activeContributions.has(contributionKey(input.slot, input.sourceId)) || context.attendedMs === 0) return;
      const key = inputRemainderKey(meter.id, inputIndex);
      const sign = input.deltaPerAttendedHour < 0 ? -1 : 1;
      const [steps, remainder] = wholeSteps(Math.abs(input.deltaPerAttendedHour), context.attendedMs, state.inputRemainders[key]!);
      state.inputRemainders[key] = remainder;
      rateDelta += sign * steps;
    });
    value = clampValue(value + factDelta + rateDelta);
    state.values[meter.id] = value;
    const fromBand = bandFor(meter, before);
    const toBand = bandFor(meter, value);
    if (fromBand !== toBand) changes.push(Object.freeze({ meterId: meter.id, fromBand, toBand, direction: value > before ? "up" : "down", valueBefore: before, valueAfter: value }));
  }
  return Object.freeze(changes);
}

export function bandFor(meter: MeterDefinition, value: number): string {
  let band = meter.bands[0]!.id;
  for (const candidate of meter.bands.slice(1)) {
    if (value < candidate.floorValue) break;
    band = candidate.id;
  }
  return band;
}

function wholeSteps(rate: number, elapsedMs: number, remainder: number): readonly [number, number] {
  const numerator = rate * elapsedMs + remainder;
  if (!Number.isSafeInteger(numerator)) invalidState();
  return [Math.floor(numerator / MILLIS_PER_HOUR), numerator % MILLIS_PER_HOUR];
}

function clampValue(value: number): number {
  return Math.max(METER_MIN_VALUE, Math.min(METER_MAX_VALUE, value));
}

function integerInRange(value: unknown, minimum: number, maximum: number): value is number {
  return Number.isSafeInteger(value) && (value as number) >= minimum && (value as number) <= maximum;
}

function sameStrings(left: readonly string[], right: readonly string[]): boolean {
  return left.length === right.length && left.every((value, index) => value === right[index]);
}

function invalidState(): never {
  throw new RangeError("invalid meter state");
}

File: server/meters/transition.go
package meters

import (
	"errors"
	"fmt"
)

const (
	MillisPerHour       int64 = 3_600_000
	MaximumAttendedStep int64 = 86_400_000
)

var ErrInvalidState = errors.New("invalid meter state")

type State struct {
	Values          map[string]int
	DecayRemainders map[string]int64
	InputRemainders map[string]int64
}

type AdvanceContext struct {
	AttendedMS          int64
	NewFactKinds        map[string]bool
	ActiveContributions map[string]bool
}

type BandChange struct {
	MeterID     string `json:"meter_id"`
	FromBand    string `json:"from_band"`
	ToBand      string `json:"to_band"`
	Direction   string `json:"direction"`
	ValueBefore int    `json:"value_before"`
	ValueAfter  int    `json:"value_after"`
}

func InputRemainderKey(meterID string, inputIndex int) string {
	return fmt.Sprintf("%s:%d", meterID, inputIndex)
}

func ContributionKey(slot, sourceID string) string {
	return slot + "\x00" + sourceID
}

func NewState(catalog *Catalog) State {
	state := State{Values: map[string]int{}, DecayRemainders: map[string]int64{}, InputRemainders: map[string]int64{}}
	if catalog == nil {
		return state
	}
	for _, meter := range catalog.Meters {
		state.Values[meter.ID] = meter.InitialValue
		state.DecayRemainders[meter.ID] = 0
		for index, input := range meter.Inputs {
			if input.Kind == InputContributionSlot {
				state.InputRemainders[InputRemainderKey(meter.ID, index)] = 0
			}
		}
	}
	return state
}

func ValidateState(catalog *Catalog, state State) error {
	if catalog == nil || state.Values == nil || state.DecayRemainders == nil || state.InputRemainders == nil {
		return ErrInvalidState
	}
	expectedInputs := map[string]bool{}
	if len(state.Values) != len(catalog.Meters) || len(state.DecayRemainders) != len(catalog.Meters) {
		return ErrInvalidState
	}
	for _, meter := range catalog.Meters {
		value, valueOK := state.Values[meter.ID]
		remainder, remainderOK := state.DecayRemainders[meter.ID]
		if !valueOK || !remainderOK || value < MinimumValue || value > MaximumValue || remainder < 0 || remainder >= MillisPerHour {
			return ErrInvalidState
		}
		for index, input := range meter.Inputs {
			if input.Kind != InputContributionSlot {
				continue
			}
			key := InputRemainderKey(meter.ID, index)
			expectedInputs[key] = true
			if remainder, ok := state.InputRemainders[key]; !ok || remainder < 0 || remainder >= MillisPerHour {
				return ErrInvalidState
			}
		}
	}
	if len(expectedInputs) != len(state.InputRemainders) {
		return ErrInvalidState
	}
	return nil
}

func Advance(catalog *Catalog, state State, context AdvanceContext) ([]BandChange, error) {
	if context.AttendedMS < 0 || context.AttendedMS > MaximumAttendedStep || context.NewFactKinds == nil || context.ActiveContributions == nil {
		return nil, ErrInvalidState
	}
	if err := ValidateState(catalog, state); err != nil {
		return nil, err
	}
	changes := make([]BandChange, 0)
	for _, meter := range catalog.Meters {
		before := state.Values[meter.ID]
		value := before
		if meter.Decay != nil && value == meter.Decay.TowardValue {
			state.DecayRemainders[meter.ID] = 0
		} else if meter.Decay != nil && context.AttendedMS > 0 {
			steps, remainder := wholeSteps(int64(meter.Decay.RatePerHour), context.AttendedMS, state.DecayRemainders[meter.ID])
			if value < meter.Decay.TowardValue {
				value += int(steps)
				if value >= meter.Decay.TowardValue {
					value, remainder = meter.Decay.TowardValue, 0
				}
			} else {
				value -= int(steps)
				if value <= meter.Decay.TowardValue {
					value, remainder = meter.Decay.TowardValue, 0
				}
			}
			state.DecayRemainders[meter.ID] = remainder
		}
		factDelta, rateDelta := 0, int64(0)
		for inputIndex, input := range meter.Inputs {
			switch input.Kind {
			case InputLedgerFact:
				if context.NewFactKinds[input.FactKind] {
					factDelta += input.Delta
				}
			case InputContributionSlot:
				if !context.ActiveContributions[ContributionKey(string(input.Slot), input.SourceID)] || context.AttendedMS == 0 {
					continue
				}
				key := InputRemainderKey(meter.ID, inputIndex)
				magnitude := int64(input.DeltaPerAttendedHour)
				sign := int64(1)
				if magnitude < 0 {
					magnitude, sign = -magnitude, -1
				}
				steps, remainder := wholeSteps(magnitude, context.AttendedMS, state.InputRemainders[key])
				state.InputRemainders[key] = remainder
				rateDelta += sign * steps
			}
		}
		value = clampValue(int64(value) + int64(factDelta) + rateDelta)
		state.Values[meter.ID] = value
		from, to := BandFor(meter, before), BandFor(meter, value)
		if from != to {
			direction := "down"
			if value > before {
				direction = "up"
			}
			changes = append(changes, BandChange{MeterID: meter.ID, FromBand: from, ToBand: to, Direction: direction, ValueBefore: before, ValueAfter: value})
		}
	}
	return changes, nil
}

func wholeSteps(rate, elapsedMS, remainder int64) (int64, int64) {
	numerator := rate*elapsedMS + remainder
	return numerator / MillisPerHour, numerator % MillisPerHour
}

func clampValue(value int64) int {
	if value < MinimumValue {
		return MinimumValue
	}
	if value > MaximumValue {
		return MaximumValue
	}
	return int(value)
}

func BandFor(meter Meter, value int) string {
	band := meter.Bands[0].ID
	for _, candidate := range meter.Bands[1:] {
		if value < candidate.FloorValue {
			break
		}
		band = candidate.ID
	}
	return band
}
